Fix for XDG_DATA_DIRS override

ruanformigoni opened the pull request:
This avoids all snaps to disappear when flatpak is installed in ubuntu 20.04. Related to issue #3678.

ruanformigoni commented:
> In `env.d/60-flatpak` we put the flatpak directories before the previous value of `$XDG_DATA_DIRS` not after, and that probably makes sense, since the act of installing the flatpak package means you want them to show up, perhaps even in case of a conflict.

Thanks for answering, I used a fresh ubuntu 20.04 install, installed some snap apps. Afterwards, I installed flatpak following the tutorial in [here](https://flatpak.org/setup/Ubuntu/), step-by-step. When I rebooted my machine, all snaps where gone.
The only solution to get both `snap` and `flatpak` working together was to modify:
```
XDG_DATA_DIRS=$HOME/.local/share/flatpak/exports/share/:/var/lib/flatpak/exports/share/:/usr/local/share/:/usr/share/
```
to
```
XDG_DATA_DIRS=$XDG_DATA_DIRS:$HOME/.local/share/flatpak/exports/share/:/var/lib/flatpak/exports/share/:/usr/local/share/:/usr/share/
```

mwleeds commented:
Are you saying you tried the following:
```
XDG_DATA_DIRS=$HOME/.local/share/flatpak/exports/share/:/var/lib/flatpak/exports/share/:/usr/local/share/:/usr/share/:$XDG_DATA_DIRS
```
and it did not allow snaps and flatpaks to both work?

I wonder if it should actually be
```
XDG_DATA_DIRS=$HOME/.local/share/flatpak/exports/share/:/var/lib/flatpak/exports/share/:$XDG_DATA_DIRS
```
because `$XDG_DATA_DIRS` should already contain `/usr/share/` and `/usr/local/share/`.

alexlarsson commented:
It is not true that `$XDG_DATA_DIRS` normally contains /usr/share. The [spec](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) says:
>  If $XDG_DATA_DIRS is either not set or empty, a value equal to /usr/local/share/:/usr/share/ should be used. 
So, if we set the env var, we should add these, or we break those.

Also, I think the patch is broken in the case where $XDG_DATA_DIR is previously unset, as that will expand to an extra ":" at the start.

alexlarsson commented:
I'm not sure exactly what format the gdm env.d directory supports for optionally appending to env vars though. @halfline ideas?

alexlarsson commented:
I guess with all the other things we do these days, such as the /usr/lib/systemd/user-environment-generators/60-flatpak generators, we should maybe just drop the gdm env.d file?

smcv commented:
If it was a shell script you'd want to be using the `${XDG_DATA_DIRS:-/usr/local/share:/usr/share}` syntax, but I don't know whether GDM env.d supports that.

systemd environment generators delegate the supported syntax to whatever language you use to write the generator, and the `60-flatpak` systemd environment generator is a shell script that uses exactly that syntax.

The `environment.d(5)` format (which originates from systemd, but is service-manager-agnostic and would be reimplementable by people who don't like systemd) is not a shell script, but has specific support for `${x:-y}` and `${x:+y}`.

smcv commented:
@ruanformigoni, if you delete `/usr/share/gdm/env.d/flatpak.env` and reboot, do both Flatpak and Snap apps show up?

ruanformigoni commented:
> @ruanformigoni, if you delete `/usr/share/gdm/env.d/flatpak.env` and reboot, do both Flatpak and Snap apps show up?

I'll try to replicate the issue in a VM, remove `/usr/share/gdm/env.d/flatpak.env` and post a video of the process.

halfline commented:
alex, you wrote the gdm env.d patch :-)  it doesn't support appending.  I say just drop it, since systemd environment files can handle it now.

refi64 commented:
For the record, I don't think environment files can append either:
https://github.com/systemd/systemd/issues/9025

On Mon, Jun 15, 2020, 9:06 AM Ray Strode <notifications@github.com> wrote:

> alex, you wrote the gdm env.d patch :-) it doesn't support appending. I
> say just drop it, since systemd environment files can handle it now.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flatpak/flatpak/pull/3679#issuecomment-644156481>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAM4YSPSNLM5WWKVKQYDIKTRWYTHVANCNFSM4N3S2JLQ>
> .
>

smcv commented:
systemd *units* can't append to search paths.

I hope environment.d(5) can, otherwise this (on Debian) is wrong:

```
$ cat /etc/environment.d/90atk-adaptor.conf
GTK_MODULES=${GTK_MODULES:+$GTK_MODULES:}gail:atk-bridge
```

halfline commented:
yea system.d can definitely do it. I know because I wrote the feature :-)

https://github.com/systemd/systemd/commit/b82f58bfe396b395bce3452bc0ba2f972fb01ab8

alexlarsson commented:
I'm thinking we can drop installing it by default, but ship it as a "if you need it" kind of thing.

alexlarsson commented:
> alex, you wrote the gdm env.d patch :-) it doesn't support appending. I say just drop it, since systemd environment files can handle it now.

Yeah, i was vaguely aware of that, but that could have changed since then.

alexlarsson commented:
I made https://github.com/flatpak/flatpak/pull/3693 instead

alexlarsson commented:
@ruanformigoni Did you try if removing the gdm file worked?

ruanformigoni commented:
> @ruanformigoni Did you try if removing the gdm file worked?

* I changed `flatpak.env` back to `XDG_DATA_DIRS=$HOME/.local/share/flatpak/exports/share/:/var/lib/flatpak/exports/share/:/usr/local/share/:/usr/share/` and rebooted.

* The snaps disappeared.

* I removed `/usr/share/gdm/env.d/flatpak.env`, and gdm did not login anymore. So I entered another tty and restored the file, and was able to login again.

* Lastly I changed `flatpak.env` back to `XDG_DATA_DIRS=$XDG_DATA_DIRS:$HOME/.local/share/flatpak/exports/share/:@localstatedir@/lib/flatpak/exports/share/:/usr/local/share/:/usr/share/` and the snaps showed up again.

alexlarsson commented:
Its very strange that removing the env file broke gdm?? I don't see how that could happen. That file is not there if you e.g. don't install flatpak...

ruanformigoni commented:
> Its very strange that removing the env file broke gdm?? I don't see how that could happen. That file is not there if you e.g. don't install flatpak...

If I uninstall flatpak correctly, gdm still works fine and `flatpak.env` disappears.

alexlarsson commented:
> > Its very strange that removing the env file broke gdm?? I don't see how that could happen. That file is not there if you e.g. don't install flatpak...
> 
> If I uninstall flatpak correctly, gdm still works fine and `flatpak.env` disappears.

But if tlatpak is installed, without the file, then gdm breaks??? Thats very weird.

alexlarsson commented:
I merged #3693 instead, will make sure it works well when I update the ppa.
